packages/storage-plus/src/keys.rs

```diff
@@ -12,6 +12,7 @@ pub enum Key<'a> {
     Val32([u8; 4]),
     Val64([u8; 8]),
     Val128([u8; 16]),
+    Bytes(Box<Vec<u8>>),
```

Comment:
This is not needed, as you can use `Vec<u8>`, or even `&[u8]` as a keys directly. This is becase `PrimaryKey` is implemented for those types.

Reply:
My intention is to make it more convenient to impl `Key`. Instead of pre allocate `Vec<u8>` and use that as `Key`, we can create any custom temporary `Key` from an arbitrary function.

The primary example is serializing/deserializing the enum variant to/from its `Key` form. We can't just use `Key::Ref` because the bytes representation of the `AssetInfo` is created temporarily in the function. (Code below)

```rust
fn key(&self) -> Vec<Key> {
    vec![Key::Ref(
        match self {
            AssetInfo::Native(native) => format!("n-{}", native),
            AssetInfo::Cw20(cw20) => format!("c-{}", cw20),
        }
        .as_bytes(),
    )]
}
```

Reply:
Instead, in order to make it work properly, we need to wrap it into an additional struct, let's say `AssetInfoKey`, and then serialize/deserialize through it. Which imo is pretty redundant.

```rust
struct AssetInfoKey(Vec<u8>);
```

Reply:
This can be solve using `Key::Bytes` above, since `Box<Vec<u8>>` allocated here is dropped at the end of the scope anyway.

```rust
fn key(&self) -> Vec<Key> {
    vec![Key::Bytes(Box::new(
        match self {
            AssetInfo::Native(native) => format!("n-{}", native),
            AssetInfo::Cw20(cw20) => format!("c-{}", cw20),
        }
        .as_bytes()
        .to_vec(),
    ))]
}
```

Reply:
> My intention is to make it more convenient to impl `Key`. Instead of pre allocate `Vec<u8>` and use that as `Key`, we can create any custom temporary `Key` from an arbitrary function.

Yes, but the decision not to do that is deliberate. For performance and efficiency reasons, we want the `key` method to return a reference. 

We only hesitantly introduced the `Key` enum and its `Val` variants, because the internal representation of integer types does not match their lexicographical sorting order (basically, because wasm32 ints repr is little-endian).
So, `Val` is there on the condition of the values to be small. Up to 16 bytes max.

Adding a generic `Bytes` will defeat that, so, we will not do it.

Reply:
Thanks for clarifying! Then these changes will not be needed!